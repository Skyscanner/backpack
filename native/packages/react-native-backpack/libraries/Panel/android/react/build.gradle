apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
