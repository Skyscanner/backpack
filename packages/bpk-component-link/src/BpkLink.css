@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-link{padding:0;border:0;background-color:transparent;text-decoration:none;cursor:pointer;appearance:none;color:#0770e3;color:var(--bpk-link-color, #0770e3)}.bpk-no-touch-support .bpk-link:hover:not(:active):not(:disabled){text-decoration:underline;color:#084eb2;color:var(--bpk-link-hover-color, #084eb2)}:global(.bpk-no-touch-support) .bpk-link:hover:not(:active):not(:disabled){text-decoration:underline;color:#084eb2;color:var(--bpk-link-hover-color, #084eb2)}.bpk-link:visited{color:#0770e3;color:var(--bpk-link-visited-color, #0770e3)}.bpk-link:active{text-decoration:underline;color:#042759;color:var(--bpk-link-active-color, #042759)}.bpk-link--active{color:#042759}.bpk-link--active:visited{color:#042759}.bpk-link--active:active{color:#042759}.bpk-link--alternate{color:#fff;color:var(--bpk-link-alternate-color, #fff)}.bpk-no-touch-support .bpk-link--alternate:hover:not(:active):not(:disabled){color:#fff;color:var(--bpk-link-alternate-hover-color, #fff)}:global(.bpk-no-touch-support) .bpk-link--alternate:hover:not(:active):not(:disabled){color:#fff;color:var(--bpk-link-alternate-hover-color, #fff)}.bpk-link--alternate:visited{color:#fff;color:var(--bpk-link-alternate-visited-color, #fff)}.bpk-link--alternate:active{color:#dddde5;color:var(--bpk-link-alternate-active-color, #dddde5)}
