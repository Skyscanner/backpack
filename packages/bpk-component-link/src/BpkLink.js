/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';

import { cssModules } from '../../bpk-react-utils';

import themeAttributes, {
  linkAlternateThemeAttributes,
} from './themeAttributes';
import STYLES from './BpkLink.module.scss';

const getClassName = cssModules(STYLES);

type Props = {
  children: Node,
  href: ?string,
  className: ?string,
  onClick: ?(event: SyntheticEvent<>) => mixed,
  blank: boolean,
  rel: ?string,
  alternate: boolean,
  white: boolean,
};

const BpkLink = (props: Props) => {
  const {
    alternate,
    blank,
    children,
    className,
    href,
    onClick,
    rel: propRel,
    white,
    ...rest
  } = props;

  const classNames = [getClassName('bpk-link')];

  const target = blank ? '_blank' : null;
  const rel = blank ? propRel || 'noopener noreferrer' : propRel;

  if (white || alternate) {
    classNames.push(getClassName('bpk-link--alternate'));
  }
  if (className) {
    classNames.push(className);
  }

  return (
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    <a
      className={classNames.join(' ')}
      href={href}
      onClick={onClick}
      target={target}
      rel={rel}
      {...rest}
    >
      {children}
    </a>
  );
};

BpkLink.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  href: PropTypes.string.isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  blank: PropTypes.bool,
  rel: PropTypes.string,
  alternate: PropTypes.bool,
  // DEPRECATED
  white: PropTypes.bool,
};

BpkLink.defaultProps = {
  className: null,
  onClick: null,
  blank: false,
  rel: null,
  alternate: false,
  // DEPRECATED
  white: false,
};

export default BpkLink;
export { themeAttributes, linkAlternateThemeAttributes };
