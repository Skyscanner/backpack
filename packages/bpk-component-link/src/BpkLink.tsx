/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */
import PropTypes from 'prop-types';
// @ts-expect-error TS(2305): Module '"react"' has no exported member 'Node'.
import type { Node } from 'react';
import { forwardRef } from 'react';

import { cssModules } from '../../bpk-react-utils';

import themeAttributes, {
  linkAlternateThemeAttributes,
} from './themeAttributes';

// @ts-expect-error TS(2307): Cannot find module './BpkLink.module.scss' or its ... Remove this comment to see the full error message
import STYLES from './BpkLink.module.scss';

const getClassName = cssModules(STYLES);

type Props = {
  children: Node,
  // @ts-expect-error TS(8020): JSDoc types can only be used inside documentation ... Remove this comment to see the full error message
  href: ?string,
  // @ts-expect-error TS(8020): JSDoc types can only be used inside documentation ... Remove this comment to see the full error message
  className: ?string,
  // @ts-expect-error TS(8020): JSDoc types can only be used inside documentation ... Remove this comment to see the full error message
  onClick: ?(event: SyntheticEvent<>) => mixed,
  blank: boolean,
  // @ts-expect-error TS(8020): JSDoc types can only be used inside documentation ... Remove this comment to see the full error message
  rel: ?string,
  alternate: boolean,
  implicit: boolean,
};

const BpkLink = forwardRef((props: Props, ref) => {
  const {
    alternate,
    blank,
    children,
    className,
    href,
    implicit,
    onClick,
    rel: propRel,
    ...rest
  } = props;

  const classNames = [getClassName('bpk-link')];
  const underlinedClassNames = [getClassName('bpk-link-underlined')];

  const target = blank ? '_blank' : null;
  const rel = blank ? propRel || 'noopener noreferrer' : propRel;

  if (className) {
    classNames.push(className);
  }
  if (implicit) {
    classNames.push(getClassName('bpk-link--implicit'));
  }
  if (alternate) {
    classNames.push(getClassName('bpk-link--alternate'));
  }
  
  if (implicit && !alternate) {
    underlinedClassNames.push(getClassName('bpk-link-underlined--implicit'));
  } else if (alternate && !implicit) {
    underlinedClassNames.push(getClassName('bpk-link-underlined--alternate'));
  } else if (implicit && alternate) {
    underlinedClassNames.push(getClassName('bpk-link-underlined-implicit--alternate'));
  }

  return (
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    <a
      className={classNames.join(' ')}
      // @ts-expect-error TS(2322): Type 'string | null' is not assignable to type 'st... Remove this comment to see the full error message
      href={href}
      // @ts-expect-error TS(2322): Type '((event: SyntheticEvent) => mixed) | null' i... Remove this comment to see the full error message
      onClick={onClick}
      // @ts-expect-error TS(2322): Type '"_blank" | null' is not assignable to type '... Remove this comment to see the full error message
      target={target}
      // @ts-expect-error TS(2322): Type 'string | null' is not assignable to type 'st... Remove this comment to see the full error message
      rel={rel}
      // @ts-expect-error TS(2322): Type 'ForwardedRef<unknown>' is not assignable to ... Remove this comment to see the full error message
      ref={ref}
      {...rest}
    >
      <span className={underlinedClassNames.join(' ')}>{children}</span>
    </a>
  );
});

BpkLink.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  href: PropTypes.string.isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  // @ts-expect-error TS(2322): Type 'Requireable<boolean>' is not assignable to t... Remove this comment to see the full error message
  blank: PropTypes.bool,
  rel: PropTypes.string,
  // @ts-expect-error TS(2322): Type 'Requireable<boolean>' is not assignable to t... Remove this comment to see the full error message
  alternate: PropTypes.bool,
  // @ts-expect-error TS(2322): Type 'Requireable<boolean>' is not assignable to t... Remove this comment to see the full error message
  implicit: PropTypes.bool,
};

BpkLink.defaultProps = {
  className: null,
  onClick: null,
  blank: false,
  rel: null,
  alternate: false,
  implicit: false,
};

export default BpkLink;
export { themeAttributes, linkAlternateThemeAttributes };
