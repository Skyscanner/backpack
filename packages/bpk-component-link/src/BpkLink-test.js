/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import { render } from '@testing-library/react';

import BpkLink, { themeAttributes } from './BpkLink';

describe('BpkLink', () => {
  it('should render correctly with a "href" attribute', () => {
    const { asFragment } = render(<BpkLink href="#">Link</BpkLink>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with a "className" attribute', () => {
    const { asFragment } = render(
      <BpkLink href="#" className="test-class">
        Link
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with a "blank" attribute', () => {
    const { asFragment } = render(
      <BpkLink href="#" blank>
        Link (new window)
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with a "rel" attribute', () => {
    const { asFragment } = render(
      <BpkLink href="#" blank rel="rel-attr">
        Link (new window)
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with "blank" and "rel" attributes', () => {
    const { asFragment } = render(
      <BpkLink href="#" blank rel="rel-overwrite">
        Link (new window)
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with a "alternate" attribute', () => {
    const { asFragment } = render(
      <BpkLink href="#" alternate>
        Link
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with arbitrary attributes', () => {
    const { asFragment } = render(
      <BpkLink href="#" id="test-id">
        Link
      </BpkLink>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('themeAttributes', () => {
    it('exports the expected themeAttributes', () => {
      expect(themeAttributes).toEqual([
        'linkColor',
        'linkHoverColor',
        'linkActiveColor',
        'linkVisitedColor',
      ]);
    });
  });
});
