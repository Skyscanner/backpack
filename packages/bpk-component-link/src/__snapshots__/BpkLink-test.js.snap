// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkLink should render correctly with "blank" and "rel" attributes 1`] = `
<DocumentFragment>
  <a
    class="bpk-link"
    href="#"
    rel="rel-overwrite"
    target="_blank"
  >
    Link (new window)
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with a "alternate" attribute 1`] = `
<DocumentFragment>
  <a
    class="bpk-link bpk-link--alternate"
    href="#"
  >
    Link
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with a "blank" attribute 1`] = `
<DocumentFragment>
  <a
    class="bpk-link"
    href="#"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link (new window)
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with a "className" attribute 1`] = `
<DocumentFragment>
  <a
    class="bpk-link test-class"
    href="#"
  >
    Link
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with a "href" attribute 1`] = `
<DocumentFragment>
  <a
    class="bpk-link"
    href="#"
  >
    Link
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with a "rel" attribute 1`] = `
<DocumentFragment>
  <a
    class="bpk-link"
    href="#"
    rel="rel-attr"
    target="_blank"
  >
    Link (new window)
  </a>
</DocumentFragment>
`;

exports[`BpkLink should render correctly with arbitrary attributes 1`] = `
<DocumentFragment>
  <a
    class="bpk-link"
    href="#"
    id="test-id"
  >
    Link
  </a>
</DocumentFragment>
`;
